### PR TITLE
azure: Switch to Azure Resource Manager from classic manager

### LINKED ISSUE #318: azure: Fix virtual networking setup
Right now, Azure boots each machine in a completely new virtual network. This means private IP's aren't really good for anything, and all communication must happen over public IP's.

The fix for this is to create a single virtual network as part of the provider initialization, and then boot each machine inside that virtual network.

I spent some time playing around with this in the azurecli, and it seems that virtual networking may not be properly supported in the "classic manager" -- instead, we'll have to use the "azure resource manager".

### LINKED ISSUE #368: Azure: Switch to resource manager from classic manager
Using classic manager has, apparently, caused all sorts of delightful problems.  Should switch to the resource manager

### PR DESCRIPTION
**Pull request by ZhongtianWang**
Before this patch, each vm was not booted in the same virtual network, and we are not taking adventage of private IPs. In this patch, we migrated from classic manager to Azure Resource Manager, as classic manager does not support the current model well.

This should close #318, close #368

### CODE REVIEW

File: provider/azure.go [hunk 1]
```diff
@@ -1,44 +1,52 @@
...
+	storageName             string
+	storageType             string
+	subnet1Name             string
+	vnetName                string
+	vnetAddressPrefix       string
+	vnetSubnetAddressPrefix string
+	nsgName                 string
+	imagePublisher          string
+	imageOffer              string
+	imageSku                string
```
Comment (ejj) [inline provider/azure.go]: A lot of these things should be consts outside of the struct I think. Basically, if you set the value at struct initialization time, and it never changes, it should be const.

File: provider/azure.go [hunk 2]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
 		return errors.New("namespace cannot be empty")
 	}
 
-	keyfile := filepath.Join(os.Getenv("HOME"), ".azure", "azure.publishsettings")
+	var err error
```
Comment (ejj) [inline provider/azure.go]: No need to declare `err` here.  Line below can just be:
```
if err := clst.loadCredentials(); err != nil {
```

File: provider/azure.go [hunk 3]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+
+	err = clst.loadCredentials()
+	if err != nil {
+		return errors.New("failed loading credentials")
+	}
 
-	azureClient, err := management.ClientFromPublishSettingsFile(keyfile, "")
+	clst.spt, err = clst.newServicePrincipalTokenFromCredentials()
```
Comment (ejj) [inline provider/azure.go]: What a silly name for a function ... "newSericePrincipalTokenFromCredentials" ... Microsoft =(

File: provider/azure.go [hunk 4]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
 		return errors.New("namespace cannot be empty")
 	}
 
-	keyfile := filepath.Join(os.Getenv("HOME"), ".azure", "azure.publishsettings")
+	var err error
+
+	err = clst.loadCredentials()
+	if err != nil {
+		return errors.New("failed loading credentials")
```
Comment (ejj) [inline provider/azure.go]: "failed to load azure credentials"

File: provider/azure.go [hunk 5]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+	}
 
-	azureClient, err := management.ClientFromPublishSettingsFile(keyfile, "")
+	clst.spt, err = clst.newServicePrincipalTokenFromCredentials()
 	if err != nil {
-		return errors.New("error retrieving azure client from publishsettings")
+		return errors.New("failed generating service principal token")
```
Comment (ejj) [inline provider/azure.go]: "failed to generate service principal token"

File: provider/azure.go [hunk 6]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+	clst.vnetSubnetAddressPrefix = "10.0.0.0/24"
+	clst.nsgName = "quiltnsg"
+	clst.imagePublisher = "Canonical"
+	clst.imageOffer = "UbuntuServer"
+	clst.imageSku = "15.10"
+	clst.imageVersion = "15.10.201511161"
+	clst.adminUsername = "ubuntu"
+	clst.adminPassword = uuid.NewV4().String()
+
+	err = clst.createOrUpdateResourceGroup()
```
Comment (ejj) [inline provider/azure.go]: combine this and the next line

File: provider/azure.go [hunk 7]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
 			continue
 		}
-		id := hs.ServiceName
 
-		// Will return empty string if the hostedservice does not have a deployment.
-		// e.g. some hosted services contains only a storage account, but no deployment.
-		deploymentName, err := clst.vmClient.GetDeploymentName(id)
-		if err != nil {
-			return nil, err
+		tags := *vm.Tags
```
Comment (ejj) [inline provider/azure.go]:
```
tag := *vm.Tags[clst.tagKey]
if tag == nil || tag != ...
```

File: provider/azure.go [hunk 8]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+		if tags[clst.tagKey] == nil || *tags[clst.tagKey] != clst.namespace {
+			continue
 		}
 
-		if deploymentName == "" {
-			clst.instanceDel(id)
-			continue
+		nicname := *vm.Name + "-nic"
+
+		iResult, err := ic.Get(clst.rgName, nicname, "")
```
Comment (ejj) [inline provider/azure.go]: This line is a bit mysterious.  Could benefit from a short comment.

Comment (ejj) [inline provider/azure.go]: also, iResult isn't a really good name for this variable.  Not particularly descriptive.

File: provider/azure.go [hunk 9]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
-			return nil, err
+		tags := *vm.Tags
+		if tags[clst.tagKey] == nil || *tags[clst.tagKey] != clst.namespace {
+			continue
 		}
 
-		if deploymentName == "" {
-			clst.instanceDel(id)
-			continue
+		nicname := *vm.Name + "-nic"
```
Comment (ejj) [inline provider/azure.go]: Btw, why not just call `nicname` `nic`?  At the very least `nicName` is easier to read.

File: provider/azure.go [hunk 10]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
-	clst.hsClient = hostedservice.NewClient(azureClient)
-	clst.vmClient = virtualmachine.NewClient(azureClient)
 	clst.namespace = namespace
-	clst.storageAccount = storageAccount
-	clst.location = clusterLocation
-	clst.vmImage = vmImage
-	clst.username = username
-	clst.userPassword = uuid.NewV4().String() // Randomly generate pwd
+	clst.rgName = "quilt"
+	clst.rgLocation = "centralus"
```
Comment (ejj) [inline provider/azure.go]: This should be set-able btw.  Fine to do it in another patch

File: provider/azure.go [hunk 11]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+	clst.tagKey = "quilt-namespace"
+	clst.storageName = "quiltstorageaccnt"
+	clst.storageType = "Standard_GRS"
+	clst.subnet1Name = "quiltsubnet1"
```
Comment (ejj) [inline provider/azure.go]: why is this subnet1Name? Why not just subnetName? Or just subnet?

File: provider/azure.go [hunk 12]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+	clst.vnetName = "quiltvnet"
+	clst.vnetAddressPrefix = "10.0.0.0/16"
```
Comment (ejj) [inline provider/azure.go]: These two things could use comments.  Why are they not both /16?  Why not /8?

Also, if we have some choice the 172 prefix (look it up precisely), is slightly better.  Less confusing as we use 10/8 internally within Quilt.

File: provider/azure.go [hunk 13]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+	clst.vnetSubnetAddressPrefix = "10.0.0.0/24"
+	clst.nsgName = "quiltnsg"
+	clst.imagePublisher = "Canonical"
+	clst.imageOffer = "UbuntuServer"
+	clst.imageSku = "15.10"
```
Comment (ejj) [inline provider/azure.go]: btw, if Ubuntu 16.04 is available, that's slightly better.

File: provider/azure.go [hunk 14]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+		nicname := *vm.Name + "-nic"
+
+		iResult, err := ic.Get(clst.rgName, nicname, "")
+		if err != nil {
+			return nil, err
 		}
+		privateIP := *(*iResult.Properties.IPConfigurations)[0].Properties.PrivateIPAddress
 
-		deploymentResponse, err := clst.vmClient.GetDeployment(id, deploymentName)
+		pipResult, err := pipac.Get(clst.rgName, nicname, "")
```
Comment (ejj) [inline provider/azure.go]: pipResult could just be pip

File: provider/azure.go [hunk 15]
```diff
@@ -47,69 +55,94 @@ func (clst *azureCluster) Connect(namespace string) error {
...
-	clst.azureClient = azureClient
-	clst.hsClient = hostedservice.NewClient(azureClient)
-	clst.vmClient = virtualmachine.NewClient(azureClient)
 	clst.namespace = namespace
-	clst.storageAccount = storageAccount
-	clst.location = clusterLocation
-	clst.vmImage = vmImage
-	clst.username = username
-	clst.userPassword = uuid.NewV4().String() // Randomly generate pwd
+	clst.rgName = "quilt"
```
Comment (ejj) [inline provider/azure.go]: Maybe "resourceGroup"

File: provider/azure.go [hunk 16]
```diff
@@ -122,20 +155,62 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
 		panic("boot count cannot be negative")
 	}
 
+	var err error
```
Comment (ejj) [inline provider/azure.go]: Why?

File: provider/azure.go [hunk 17]
```diff
@@ -122,20 +155,62 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
...
 			return err
 		}
-	}
 
+		spf := network.SubnetPropertiesFormat{
+			AddressPrefix: &clst.vnetSubnetAddressPrefix,
+		}
+
+		subnetID := fmt.Sprintf(
+			"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s/subnets/%s",
```
Comment (ejj) [inline provider/azure.go]: This probably spills over length.  Could use a `+` on the format string.

File: provider/azure.go [hunk 18]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
 
 // Disconnect.
 func (clst *azureCluster) Disconnect() {
-	// nothing
+	gc := resources.NewGroupsClient(clst.subscriptionID)
+	gc.Authorizer = clst.spt
+	cancel := make(chan struct{})
+	gc.Delete(clst.rgName, cancel)
```
Comment (ejj) [inline provider/azure.go]: Can delete fail?  What if it fails? Does it block? How long does it block?  The fact that it takes a cancel chan is odd . . . makes me thing we aren't dealing with this function properly.

File: provider/azure.go [hunk 19]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
 
-// Create one Azure instance (blocking).
-func (clst *azureCluster) instanceNew(name string, vmSize string, region string,
-	cloudConfig string) error {
-	if region != clst.location {
-		return fmt.Errorf("cannot create instance in %s, only %s is supported", region, clusterLocation)
+func (clst *azureCluster) loadCredentials() error {
+	u, err := user.Current()
+	if err != nil {
+		return errors.New("unable to determine current OS user")
```
Comment (ejj) [inline provider/azure.go]: "current user"

File: provider/azure.go [hunk 20]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
-			ServiceName: name,
-			Description: clst.namespace,
-			Location:    clst.location,
-			Label:       base64.StdEncoding.EncodeToString([]byte(name)),
-		})
+
+	n := u.HomeDir + credentialsPath
+	f, err := os.Open(n)
+	if err != nil {
+		return errors.New("unable to open credentials at " + n)
```
Comment (ejj) [inline provider/azure.go]: "azure credentials"

File: provider/azure.go [hunk 21]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+	}
+
+	b, err := ioutil.ReadAll(f)
+	if err != nil {
+		return errors.New("unable to read " + n)
+	}
+
+	c := map[string]interface{}{}
```
Comment (ejj) [inline provider/azure.go]: Will a map[string][string] work?

Comment (ejj) [inline provider/azure.go]: Btw this whole process is a bit odd.  Are you sure that, now that we've updated the azure SDK, they haven't come up with an officially sanctioned way to pull the credentials in?  Seems like every user will have this problem ...  What do they suggest?

File: provider/azure.go [hunk 22]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+	err = json.Unmarshal(b, &c)
+	if err != nil {
+		return errors.New(n + " contained invalid JSON")
+	}
+
+	var ok bool
```
Comment (ejj) [inline provider/azure.go]: Why?

Perhaps you aren't aware of `:=`?

File: provider/azure.go [hunk 23]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
-	// create hostedservice
-	err := clst.hsClient.CreateHostedService(
-		hostedservice.CreateHostedServiceParameters{
-			ServiceName: name,
-			Description: clst.namespace,
-			Location:    clst.location,
-			Label:       base64.StdEncoding.EncodeToString([]byte(name)),
-		})
+
+	n := u.HomeDir + credentialsPath
```
Comment (ejj) [inline provider/azure.go]: `n` isn't a great name for this variable.  Sounds like a counter.

File: provider/azure.go [hunk 24]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+	}
+
+	if clst.subscriptionID, ok = c["subscriptionID"].(string); !ok {
+		return errors.New(n + " contained invalid subscriptionID")
+	}
+
+	return nil
+}
+
+func (clst *azureCluster) newServicePrincipalTokenFromCredentials() (*azure.ServicePrincipalToken, error) {
```
Comment (ejj) [inline provider/azure.go]: Thus function name is *way to long*.

File: provider/azure.go [hunk 25]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+	oauthConfig, err := azure.PublicCloud.OAuthConfigForTenant(clst.tenantID)
+	if err != nil {
+		return nil, err
+	}
+	return azure.NewServicePrincipalToken(*oauthConfig, clst.clientID, clst.clientSecret, azure.PublicCloud.ResourceManagerEndpoint)
+}
+
+func (clst *azureCluster) createOrUpdateResourceGroup() error {
```
Comment (ejj) [inline provider/azure.go]: also this one

File: provider/azure.go [hunk 26]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+		Name:     &clst.rgName,
+		Location: &clst.rgLocation,
+	}
+
+	_, err := gc.CreateOrUpdate(clst.rgName, model)
+
+	return err
+}
+
+func (clst *azureCluster) createOrUpdateStorageAccount(location string) error {
```
Comment (ejj) [inline provider/azure.go]: `configureStorageAccount`

File: provider/azure.go [hunk 27]
```diff
@@ -144,83 +219,395 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+
+	return nil
+}
+
+func (clst *azureCluster) newServicePrincipalTokenFromCredentials() (*azure.ServicePrincipalToken, error) {
+	oauthConfig, err := azure.PublicCloud.OAuthConfigForTenant(clst.tenantID)
+	if err != nil {
+		return nil, err
+	}
+	return azure.NewServicePrincipalToken(*oauthConfig, clst.clientID, clst.clientSecret, azure.PublicCloud.ResourceManagerEndpoint)
```
Comment (ejj) [inline provider/azure.go]: Line is too long.

File: provider/azure.go (hunk 1 above)
Comment (ZhongtianWang) [inline provider/azure.go]: Yeah. By the way, we should also let users specify their subscription ID, as well as OS image, right?

File: provider/azure.go (hunk 4 above)
Comment (ZhongtianWang) [inline provider/azure.go]: 😅😅😅😅😅

File: provider/azure.go (hunk 10 above)
Comment (ZhongtianWang) [inline provider/azure.go]: I thought about it too, but in the end I was thinking that the name of a resource group doesn't really matter as much, since it's just a group name that contains all kinds of resources under it (nic, vm, network interface). However, the location does impact as little bit when reading what kinds of resources does this group contain, so I definitely shouldn't just set it to central US. Should we just include all these fields in dsl? It definitely gives users more control over everything, but it might be annoying for many people, too.

File: provider/azure.go (hunk 11 above)
Comment (ZhongtianWang) [inline provider/azure.go]: I was playing with different subnets. Now I'm guessing we probably would never need more than 1 subnet, so lemme change it :P

File: provider/azure.go (hunk 21 above)
Comment (ZhongtianWang) [inline provider/azure.go]: This is so far their official way to pull in the credentials. They showed some code in an example folder, and what we currently have is very similar to what they have. The difference is, I'm checking 4 fields manually, while they use code to check all fields in a loop, and they try to convert everything to string.

File: provider/azure.go (hunk 22 above)
Comment (ZhongtianWang) [inline provider/azure.go]: haha I thought this way looks prettier in the beginning...  🙃  But `:=` should be fine

File: provider/azure.go (hunk 2 above)
Comment (ZhongtianWang) [inline provider/azure.go]: Actually if we don't declare it here, we would still need to declare it right after this line. Because we can't have something like `clst.spt, err := clst.newServicePrincipalToken()` since clst.spt is not an identifier.

File: provider/azure.go (hunk 22 above)
Comment (ZhongtianWang) [inline provider/azure.go]: Actually, we can't have line like this `if clst.clientID, ok := cred["clientID"].(string); !ok {`since clst.clientID cannot be at the left side of `:=`

**Comment by ejj** [comment]
Btw, could you rebase?  I updated the vendor directory, so I think the first patch is no longer necessary.

File: provider/azure.go [hunk 28]
```diff
@@ -1,44 +1,55 @@
...
 	uuid "github.com/satori/go.uuid"
 )
 
-const storageAccount string = "netsysstorage"
-const clusterLocation string = "Central US"
-const vmImage string = "b39f27a8b8c64d52b05eac6a62ebad85__Ubuntu-15_10-amd64-server-20151116.1-en-us-30GB"
-const username string = "ubuntu"
+const credentialsPath = "/.azure/credentials.json"
+
+var (
```
Comment (ejj) [inline provider/azure.go]: All these should be const

File: provider/azure.go [hunk 29]
```diff
@@ -1,44 +1,55 @@
...
+	"github.com/Azure/go-autorest/autorest/to"
+
 	uuid "github.com/satori/go.uuid"
 )
 
-const storageAccount string = "netsysstorage"
-const clusterLocation string = "Central US"
-const vmImage string = "b39f27a8b8c64d52b05eac6a62ebad85__Ubuntu-15_10-amd64-server-20151116.1-en-us-30GB"
-const username string = "ubuntu"
+const credentialsPath = "/.azure/credentials.json"
```
Comment (ejj) [inline provider/azure.go]: Why is this treated differently than below?

File: provider/azure.go [hunk 30]
```diff
@@ -1,44 +1,55 @@
...
+
+var (
+	tagKey                  = "quilt-namespace"
+	resourceGroup           = "quilt"
+	rgLocation              = "centralus"
+	storageName             = "quiltstorageaccnt"
```
Comment (ejj) [inline provider/azure.go]: Camel Case  Also perhaps the variable should be calle storageAccount

File: provider/azure.go [hunk 31]
```diff
@@ -1,44 +1,55 @@
...
+	storageType             = "Standard_GRS"
+	subnetName              = "quiltsubnet"
```
Comment (ejj) [inline provider/azure.go]: Just call the variable "subnet" and the value quiltSubnet  basically I don't like "Name" in variable names, it's redundant.

File: provider/azure.go [hunk 32]
```diff
@@ -1,44 +1,55 @@
...
+	vnetName                = "quiltvnet"
+	vnetAddressPrefix       = "10.0.0.0/8"
+	vnetSubnetAddressPrefix = "10.0.0.0/8"
+	nsgName                 = "quiltnsg"
```
Comment (ejj) [inline provider/azure.go]: This is pretty mysterious.  What's an nsg?  Actually come to think of it, most of these constants deserve a comment.

File: provider/azure.go [hunk 33]
```diff
@@ -1,44 +1,55 @@
...
+	clientSecret   string
+	tenantID       string
+	subscriptionID string
 	namespace      string
-	storageAccount string
-	location       string
-	vmImage        string
-	username       string
-	userPassword   string // Required password is a randomly generated UUID.
+	adminUsername  string
```
Comment (ejj) [inline provider/azure.go]: Does this need to be part of azureCluster, or can it just be generated as needed?

File: provider/azure.go (hunk 2 above)
Comment (ejj) [inline provider/azure.go]: Ah if that's the case, still cleaner to do the following 
```
err := clst.LoadCredentials
if err != nil {
    ....
}
```

File: provider/azure.go [hunk 34]
```diff
@@ -47,69 +58,79 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+	if err = clst.configureResourceGroup(); err != nil {
+		return err
+	}
 
 	return nil
 }
 
 // Retrieve list of instances.
 func (clst *azureCluster) List() ([]Machine, error) {
+	vmc := compute.NewVirtualMachinesClient(clst.subscriptionID)
```
Comment (ejj) [inline provider/azure.go]: Does it make sense to be generating this on each function call, or instead, should they be part of azureCluster?

File: provider/azure.go [hunk 35]
```diff
@@ -47,69 +58,79 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+	vmc.Authorizer = clst.spt
+
+	ic := network.NewInterfacesClient(clst.subscriptionID)
+	ic.Authorizer = clst.spt
+
+	pipac := network.NewPublicIPAddressesClient(clst.subscriptionID)
+	pipac.Authorizer = clst.spt
+
 	var mList []Machine
```
Comment (ejj) [inline provider/azure.go]: I'd define this above the for loop as that's closest to it first use.

File: provider/azure.go [hunk 36]
```diff
@@ -47,69 +58,79 @@ func (clst *azureCluster) Connect(namespace string) error {
...
-		if err != nil {
-			return nil, err
+		if tag := (*vm.Tags)[tagKey]; tag == nil || *tag != clst.namespace {
+			continue
 		}
 
-		if deploymentName == "" {
-			clst.instanceDel(id)
-			continue
+		nicName := *vm.Name + "-nic"
```
Comment (ejj) [inline provider/azure.go]: just `nic` instead of `nicName`.  Also can vm.Name be nil?  If not, why is it a *?

File: provider/azure.go [hunk 37]
```diff
@@ -47,69 +58,79 @@ func (clst *azureCluster) Connect(namespace string) error {
...
+
+		// retreive details about the specified network interface
+		ni, err := ic.Get(resourceGroup, nicName, "")
+		if err != nil {
+			return nil, err
 		}
+		privateIP := *(*ni.Properties.IPConfigurations)[0].Properties.PrivateIPAddress
```
Comment (ejj) [inline provider/azure.go]: You're doing a lot of dereferencing and slice element access.  Are you sure these things can't be nil or be empty?  Probably should check and return errors.

File: provider/azure.go [hunk 38]
```diff
@@ -47,69 +58,79 @@ func (clst *azureCluster) Connect(namespace string) error {
...
 
-		deploymentResponse, err := clst.vmClient.GetDeployment(id, deploymentName)
+		pip, err := pipac.Get(resourceGroup, nicName, "")
 		if err != nil {
 			return nil, err
 		}
+		publicIP := *pip.Properties.IPAddress
```
Comment (ejj) [inline provider/azure.go]: Could this be nil?

File: provider/azure.go [hunk 39]
```diff
@@ -123,16 +144,50 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
 	}
```
Comment (ejj) [inline provider/azure.go]: Btw, above there's some weirdness.  This panic check is useless and the function comment has a cryptic statement about instanceNew that isn't called anywhere.  Probably should fix in a separate patch.

File: provider/azure.go [hunk 40]
```diff
@@ -123,16 +144,50 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
...
 
 	for _, m := range bootSet {
-		name := "quilt-" + uuid.NewV4().String()
-		if err := clst.instanceNew(name, m.Size, m.Region, cloudConfigUbuntu(m.SSHKeys, "wily")); err != nil {
+		// One storage account is needed per region
+		if err := clst.configureStorageAccount(m.Region); err != nil {
```
Comment (ejj) [inline provider/azure.go]: Should we really be re-creating the storage account for every region _every single time_ we boot a machine?  Seems to me, this could be done once for all regions when the cluster is create.

Btw, I'm really surprised this works ... based on my reading of configureStorageAccount(), my impression is this should fail because we keep re-trying to create the storage account with the same name.

File: provider/azure.go [hunk 41]
```diff
@@ -123,16 +144,50 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
...
 			return err
 		}
-	}
 
+		spf := network.SubnetPropertiesFormat{
+			AddressPrefix: &vnetSubnetAddressPrefix,
+		}
+
+		subnetID := fmt.Sprintf(
+			"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/"+
```
Comment (ejj) [inline provider/azure.go]: This is over 89 characters.

File: provider/azure.go [hunk 42]
```diff
@@ -123,16 +144,50 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
...
+				"virtualNetworks/%s/subnets/%s",
+			clst.subscriptionID,
```
Comment (ejj) [inline provider/azure.go]: No need for a newline after each function argument.

File: provider/azure.go [hunk 43]
```diff
@@ -123,16 +144,50 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
...
+			vnetName,
+			subnetName)
+
+		subnet := network.Subnet{
+			Name:       &subnetName,
+			Properties: &spf,
+			ID:         &subnetID,
+		}
+
+		if err := clst.configureVirtualNetwork(subnet, m.Region); err != nil {
```
Comment (ejj) [inline provider/azure.go]: Again, does this really have to be done for *every* machine, or can it be done once per region?

File: provider/azure.go [hunk 44]
```diff
@@ -123,16 +144,50 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
...
+			return err
+		}
+
+		vmName := "quilt-" + uuid.NewV4().String()
+		osDiskName := vmName + "-osdisk"
+		nicName := vmName + "-nic"
+
+		if err := clst.configureNetworkInterface(nicName, subnet, m.Region); err != nil {
```
Comment (ejj) [inline provider/azure.go]: Line is too long.

Comment (ejj) [inline provider/azure.go]: btw, you don't always have to do the err check in a single if statment.  Just if the line is short.  In this case I'd just break this into two lines, one makes the function call, the other does the if check

File: provider/azure.go [hunk 45]
```diff
@@ -123,16 +144,50 @@ func (clst *azureCluster) Boot(bootSet []Machine) error {
...
+			return err
+		}
+
+		cloudConfig := cloudConfigUbuntu(m.SSHKeys, "wily")
```
Comment (ejj) [inline provider/azure.go]: Pull this outside of the loop.

File: provider/azure.go [hunk 46]
```diff
@@ -144,83 +199,399 @@ func (clst *azureCluster) Stop(machines []Machine) error {
 
 // Disconnect.
 func (clst *azureCluster) Disconnect() {
-	// nothing
+	gc := resources.NewGroupsClient(clst.subscriptionID)
+	gc.Authorizer = clst.spt
+	cancel := make(chan struct{})
+	gc.Delete(resourceGroup, cancel)
```
Comment (ejj) [inline provider/azure.go]: Does this delete all the storage accounts and virtual networks and subnets and everything we've created?

File: provider/azure.go [hunk 47]
```diff
@@ -144,83 +199,399 @@ func (clst *azureCluster) Stop(machines []Machine) error {
 
 // Disconnect.
```
Comment (ejj) [inline provider/azure.go]: This comment adds nothing btw.

File: provider/azure.go (hunk 46 above)
Comment (ejj) [inline provider/azure.go]: Actually, it doesn't really matter, we never actually call Disconnect apparently.

File: provider/azure.go [hunk 48]
```diff
@@ -144,83 +199,399 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+	if err != nil {
+		return errors.New("unable to open Azure credentials at " + dir)
+	}
+
+	b, err := ioutil.ReadAll(f)
+	if err != nil {
+		return errors.New("unable to read " + dir)
+	}
+
+	cred := map[string]interface{}{}
```
Comment (ejj) [inline provider/azure.go]: Why not map[string]string?

File: provider/azure.go [hunk 49]
```diff
@@ -144,83 +199,399 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
-	err := clst.hsClient.CreateHostedService(
-		hostedservice.CreateHostedServiceParameters{
-			ServiceName: name,
-			Description: clst.namespace,
-			Location:    clst.location,
-			Label:       base64.StdEncoding.EncodeToString([]byte(name)),
-		})
+
+	dir := u.HomeDir + credentialsPath
+	f, err := os.Open(dir)
```
Comment (ejj) [inline provider/azure.go]: Should you defer closing the file?

File: provider/azure.go [hunk 50]
```diff
@@ -144,83 +199,399 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+	}
+
+	if clst.subscriptionID, ok = cred["subscriptionID"].(string); !ok {
+		return errors.New(dir + " contains invalid subscriptionID")
+	}
+
+	return nil
+}
+
+func (clst *azureCluster) newServicePrincipalToken() (*azure.ServicePrincipalToken,
```
Comment (ejj) [inline provider/azure.go]: I think the name `newToken` is a tad cleaner.

File: provider/azure.go [hunk 51]
```diff
@@ -1,44 +1,55 @@
...
-const storageAccount string = "netsysstorage"
-const clusterLocation string = "Central US"
-const vmImage string = "b39f27a8b8c64d52b05eac6a62ebad85__Ubuntu-15_10-amd64-server-20151116.1-en-us-30GB"
-const username string = "ubuntu"
+const credentialsPath = "/.azure/credentials.json"
+
+var (
+	tagKey                  = "quilt-namespace"
+	resourceGroup           = "quilt"
+	rgLocation              = "centralus"
```
Comment (ejj) [inline provider/azure.go]: Is this going to be an issue?  How will this work with Azure VMs in other regions?

File: provider/azure.go [hunk 52]
```diff
@@ -144,83 +199,399 @@ func (clst *azureCluster) Stop(machines []Machine) error {
...
+	// Create public IP address
+	pipac := network.NewPublicIPAddressesClient(clst.subscriptionID)
+	pipac.Authorizer = clst.spt
+
+	pipapf := network.PublicIPAddressPropertiesFormat{
+		PublicIPAllocationMethod: network.Dynamic,
+	}
+
+	pipaID := fmt.Sprintf(
+		"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/"+
```
Comment (ejj) [inline provider/azure.go]: You have some very similar code strewn about to generate these IDs.  Perhaps you could pull it into a function?

**Comment by ejj** [comment]
So I'm going to stop reviewing this for now.  There are some high level issues I think need to be resolved.

1) We're doing a lot of work per-VM ... creating various subnets and what not.  I think this is likely what's causing a lot of the performance issues.

2) the SetACLs situation isn't good enough at the moment.  Basically, you're creating all of the security groups at NIC creation time, but that really should be handled dynamically in the SetACLs function.

3) Really probably should do the ChooseBestSize thingy.

At this point, this seems like a promising prototype, but until those three issues are resolved, I don't think we should merge.  Doesn't have to be in the same commit, but probably should be in PR.  I'm going to close the PR for now when that stuffs done, go ahead and resubmit.

File: provider/azure.go (hunk 28 above)
Comment (ZhongtianWang) [inline provider/azure.go]: But eventually we would need pointers to there vars. If those are const, we would still need to create temp variables and use their addresses.

File: provider/azure.go (hunk 36 above)
Comment (ZhongtianWang) [inline provider/azure.go]: Because it's a pointer

File: provider/azure.go (hunk 30 above)
Comment (ZhongtianWang) [inline provider/azure.go]: Camel case is not allowed for storage account
